src/Storages/MergeTree/GinIndexStore.h

```diff
@@ -0,0 +1,266 @@
+
+    /// Add a row_id into the builder
+    void add(UInt32 row_id);
+
+    /// Check whether the builder is using roaring bitmap
+    bool useRoaring() const;
+
+    /// Serialize the content of builder to given WriteBuffer, returns the bytes of serialized data
+    UInt64 serialize(WriteBuffer &buffer) const;
+
+    /// Deserialize the postings list data from given ReadBuffer, return a pointer to the GinIndexPostingsList created by deserialization
+    static GinIndexPostingsListPtr deserialize(ReadBuffer &buffer);
+private:
+    static constexpr UInt8 UsesBitMap = 0xFF;
+};
+
+using GinIndexPostingsBuilderPtr = std::shared_ptr<GinIndexPostingsBuilder>;
+
+/// Container for all term's Gin Index Postings List Builder
+using GinIndexPostingsBuilderContainer = std::unordered_map<std::string, GinIndexPostingsBuilderPtr>;
+
+/// Container for postings lists for each segment
+using SegmentedPostingsListContainer = std::unordered_map<UInt32, GinIndexPostingsListPtr>;
+
+/// Postings lists and terms built from query string
+using PostingsCache = std::unordered_map<std::string, SegmentedPostingsListContainer>;
+using PostingsCachePtr = std::shared_ptr<PostingsCache>;
+
+/// Term dictionary information, which contains:
+struct TermDictionary
+{
+    /// .gin_post file offset of this segment's postings lists
+    UInt64 postings_start_offset;
+
+    /// .gin_dict file offset of this segment's term dictionaries
+    UInt64 item_dict_start_offset;
+
+    /// Finite State Transducer, which can be viewed as a map of <term, offset>
+    FST::FiniteStateTransducer offsets;
+};
+
+using TermDictionaryPtr = std::shared_ptr<TermDictionary>;
+using TermDictionaries = std::unordered_map<UInt32, TermDictionaryPtr>;
+
+/// Gin Index Store which has Gin Index meta data for the corresponding Data Part
+class GinIndexStore
+{
+public:
+    explicit GinIndexStore(const String& name_, DataPartStoragePtr storage_)
+        : name(name_),
+        storage(storage_)
+    {
+    }
+    GinIndexStore(const String& name_, DataPartStoragePtr storage_, DataPartStorageBuilderPtr data_part_storage_builder_, UInt64 max_digestion_size_)
+        : name(name_),
+        storage(storage_),
+        data_part_storage_builder(data_part_storage_builder_),
+        max_digestion_size(max_digestion_size_)
+    {
+    }
+
+    bool load();
+
+    /// Check existence by checking the existence of file .gin_seg
+    bool exists() const;
+
+    UInt32 getNextIDRange(const String &file_name, size_t n);
+
+    UInt32 getNextRowIDRange(size_t n);
+
+    UInt32 getNextSegmentID();
+
+    UInt32 getSegmentNum();
+
+    using GinIndexStorePtr = std::shared_ptr<GinIndexStore>;
+
+    GinIndexPostingsBuilderContainer& getPostings() { return current_postings; }
+
+    /// Check if we need to write segment to Gin index files
+    bool needToWrite() const;
+
+    /// Accumulate the size of text data which has been digested
+    void addSize(UInt64 sz) { current_size += sz; }
```

Comment:
Since another class(`MergeTreeIndexAggregatorGinFilter`) calls this method, and also calls a related method `needsToWrite()`, it could be cumbersome if using different ways IMHO.